### PR TITLE
[1.1.x] Max7219 cleanup, simplify, and extend

### PR DESCRIPTION
**Pull request by thinkyhead**
Cleanups to Max7219 in the course of testing and debugging. No changes in core behavior.

- Repair and simplify utility macros and relocate to the `.cpp` file.
- Add `M7219 D V` command to set a native row in the matrix directly to a byte value.
- Refactor as a singleton class called `Max7219`.

### CODE REVIEW

**Comment by Roxy-3D** [comment]
Ah...   You did test all this, right?   Because I did test all four orientations.   And I'm very skeptical you can delete all this and have it still work in all four orientations.

<details><summary>Code…</summary>

```cpp
#if _ROT == 0
  #define MAX7219_UPDATE_AXIS     y   // Fast line update axis for this orientation of the matrix display
  #define MAX7219_X_LEDS          (8 * MAX7219_NUMBER_UNITS)
  #define MAX7219_Y_LEDS          8
  #define XOR_7219(x, y)          LEDs[y + (x >> 3) * 8] ^= _BV(7 - (x & 0x07))
  #define SET_PIXEL_7219(x, y)    LEDs[y + (x >> 3) * 8] |= _BV(7 - (x & 0x07))
  #define CLEAR_PIXEL_7219(x, y)  LEDs[y + (x >> 3) * 8] &= (_BV(7 - (x & 0x07)) ^ 0xFF)
  #define BIT_7219(x, y)          TEST(LEDs[y + (x >> 3) * 8], 7 - (x & 0x07))
  #define SEND_7219(R) do{ for (int8_t jj = 0; jj < MAX7219_NUMBER_UNITS; jj++) Max7219(max7219_reg_digit0 + (R & 0x7), LEDs[(R & 0x7) + jj * 8]); Max7219_pulse_load(); }while(0)
#elif _ROT == 90
  #define MAX7219_UPDATE_AXIS     x   // Fast line update axis for this orientation of the matrix display
  #define MAX7219_X_LEDS          8
  #define MAX7219_Y_LEDS          (8 * MAX7219_NUMBER_UNITS)
  #define XOR_7219(x, y)          LEDs[x + (((MAX7219_Y_LEDS - 1 - y) >> 3) * 8)] ^= _BV((y & 0x7))
  #define SET_PIXEL_7219(x, y)    LEDs[x + (((MAX7219_Y_LEDS - 1 - y) >> 3) * 8)] |= _BV((y & 0x7))
  #define CLEAR_PIXEL_7219(x, y)  LEDs[x + (((MAX7219_Y_LEDS - 1 - y) >> 3) * 8)] &= (_BV((y & 0x7)) ^ 0xFF)
  #define BIT_7219(x, y)          TEST(LEDs[x + (((MAX7219_Y_LEDS - 1 - y) >> 3) * 8)], (y & 0x7))
  #define SEND_7219(R) do{ for (int8_t jj = 0; jj < MAX7219_NUMBER_UNITS; jj++) Max7219(max7219_reg_digit0 + (R & 0x7), LEDs[(R & 0x7) + jj * 8]); Max7219_pulse_load(); }while(0)
#elif _ROT == 180
  #define MAX7219_UPDATE_AXIS     y   // Fast line update axis for this orientation of the matrix display
  #define MAX7219_X_LEDS          (8 * MAX7219_NUMBER_UNITS)
  #define MAX7219_Y_LEDS          8
  #define XOR_7219(x, y)          LEDs[y + ((MAX7219_X_LEDS - 1 - x) >> 3) * 8] ^= _BV((x & 0x07))
  #define SET_PIXEL_7219(x, y)    LEDs[y + ((MAX7219_X_LEDS - 1 - x) >> 3) * 8] |= _BV((x & 0x07))
  #define CLEAR_PIXEL_7219(x, y)  LEDs[y + ((MAX7219_X_LEDS - 1 - x) >> 3) * 8] &= (_BV((x & 0x07)) ^ 0xFF)
  #define BIT_7219(x, y)          TEST(LEDs[y + ((MAX7219_X_LEDS - 1 - x) >> 3) * 8], ((x & 0x07)))
  #define SEND_7219(R) do{ for (int8_t jj = 0; jj < MAX7219_NUMBER_UNITS; jj++) Max7219(max7219_reg_digit7 - (R & 0x7), LEDs[(R & 0x7) + jj * 8]); Max7219_pulse_load(); }while(0)
#elif _ROT == 270
  #define MAX7219_UPDATE_AXIS     x   // Fast line update axis for this orientation of the matrix display
  #define MAX7219_X_LEDS          8
  #define MAX7219_Y_LEDS          (8 * MAX7219_NUMBER_UNITS)
  #define XOR_7219(x, y)          LEDs[x + (y >> 3) * 8] ^= _BV(7 - (y & 0x7))
  #define SET_PIXEL_7219(x, y)    LEDs[x + (y >> 3) * 8] |= _BV(7 - (y & 0x7))
  #define CLEAR_PIXEL_7219(x, y)  LEDs[x + (y >> 3) * 8] &= (_BV(7 - (y & 0x7)) ^ 0xFF)
  #define BIT_7219(x, y)          TEST(LEDs[x + ( y >> 3) * 8], 7 - (y & 0x7))
  #define SEND_7219(R) do{ for (int8_t jj = 0; jj < MAX7219_NUMBER_UNITS; jj++) Max7219(max7219_reg_digit7 - (R & 0x7), LEDs[(R & 0x7) + jj * 8]); Max7219_pulse_load(); }while(0)



#define MAX7219_COLS 8
#define MAX7219_ROWS (8 * (MAX7219_NUMBER_UNITS))
#if _ROT == 0 || _ROT == 180
  #define MAX7219_X_LEDS          MAX7219_ROWS
  #define MAX7219_Y_LEDS          MAX7219_COLS
#else
  #error "MAX7219_ROTATE must be a multiple of +/- 90°."
  #define MAX7219_X_LEDS          MAX7219_COLS
  #define MAX7219_Y_LEDS          MAX7219_ROWS
#endif
```
</details>

**Comment by thinkyhead** [comment]
Yep. I'm testing as I go. Mainly in the 2.0.x version. Those macros were simply moved to the `.cpp` file to save a little memory in the compilation process.

File: Marlin/Max7219_Debug_LEDs.cpp [hunk 1]
```diff
@@ -48,7 +48,36 @@
...
+  #define MAX_IND(X,Y)  _MAX_IND(X,Y)
+  #define MAX_BIT(X,Y)  _MAX_BIT(Y)
+#else
+  #error "MAX7219_ROTATE must be a multiple of +/- 90°."
+#endif
+
+#define XOR_7219(X,Y)         LEDs[MAX_IND(X,Y)] ^= _BV(MAX_BIT(X,Y))
+#define SET_PIXEL_7219(X,Y)   LEDs[MAX_IND(X,Y)] |= _BV(MAX_BIT(X,Y))
+#define CLEAR_PIXEL_7219(X,Y) LEDs[MAX_IND(X,Y)] &= ~_BV(MAX_BIT(X,Y))
+#define BIT_7219(X,Y)         TEST(LEDs[MAX_IND(X,Y)], MAX_BIT(X,Y))
```
Comment (thinkyhead) [inline Marlin/Max7219_Debug_LEDs.cpp]: ☝️ New location of those utility macros…

**Comment by thinkyhead** [comment]
Here's a nice optimization. You can skip the units you don't care about updating and just update one. The key is to send `max7219_reg_noop` to skip a unit. According to the data sheet, when you cascade them, the last one in series is actually the first one in the logical chain. So, to only update the first unit in the series, you have to send `noop`, then send the data for the first unit. This explains why "Row 0" in my little set of two is the first row on the _second unit_ in series, and "Row 8" is the first row on the first unit in series.

So this method makes it so rows 0-7 go to the first unit, rows 8-15 go the second, etc. :

```cpp
// Send out a single native row of bits to just one unit
void Max7219::one(const uint8_t line) {
  for (uint8_t u = MAX7219_NUMBER_UNITS; u--;) {
    if (u == (line >> 3))
      send(_ROW_REG(line), led_line[line]);
    else
      noop();
  }
  pulse_load();
}
```

**Comment by Roxy-3D** [comment]
> Here's a nice optimization. You can skip the units you don't care about updating and just update one. The key is to send max7219_reg_noop to skip a unit.

Yeah...   I debated doing that.   The reason I didn't do that was because the NOOP command is just as expensive as actually sending the row of bits out.   You still need to clock out the NOOP command, and then 8 more bits of meaningless data.

So...   I was thinking, it just made sense to always update the same row of each display since we are stuck clocking out 16 bits (NOOP and 8 data bits) any way.

**Comment by thinkyhead** [comment]
> it just made sense to always update the same row of each display since we are stuck clocking out 16 bits (NOOP and 8 data bits) any way.

Yeah, at first reading of the datasheet it looked like noop was only a single byte, but yup, this chip always needs 16 bits per unit.

This PR should preserve all the existing behavior. So check and see if `MAX7219_ROTATE` set to 0 is doing what it should for your chained devices. With mine (two standalone boards wired together) it's still backward, with bits 0-7 showing on the first display in the chain and the bits 8-15 on the second.

**Pull request merged**